---
title: "TWAS analysis"
output: html_document
date: "2025-9-18"
editor_options: 
  chunk_output_type: console
---

```{r message=F}
library(ggplot2)
library(ggpubr)

library(dplyr)
library(reshape2)

options(warn = 1) # Warnings are printed immediately (some packages delay warning printing).
```

```{r}
setwd("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/")
```

# 1. Read results files

#### Brain Catalog
```{r}
brainCatalog = readxl::read_xlsx("/mnt/Data1/HwangLab/taeyoungh/miRNA/Adult_Freeze-040625/1_Dataset/BrainCatalog_Metadata.xlsx")

table(brainCatalog$ancestry) # all European
table(brainCatalog$trait_type) # binary # 123
```

Select traits
```{r}
brainCatalog = subset(brainCatalog, trait_type == "binary")
```

> we will use binary trait only

#### TWAS output
check folder name: "Fusion associations and SMR were completed for CAUC, while SMR was cancelled for AA. " 1/15/25 Arun

Fusion
```{r}
fusion.res = list()

for (cur.dir in c("DLPFC-CAUC", "mPFC-CAUC", "Hippo-CAUC", "Caudate-CAUC")) {
  cur.files = list.files(path = paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/Fusion/associations_p01/"), pattern = "_associations.txt$") %>% data.frame(filename=.) %>% mutate(ID = gsub("_associations.txt", ".txt", filename)) %>% filter(ID %in% brainCatalog$files_names)
  
  for (f in cur.files$filename) {
    print(paste(cur.dir, f))
    temp = read.table(paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/Fusion/associations_p01/", f), header = T, sep = "\t")
    fusion.res[[paste0(f, cur.dir)]] = temp %>% mutate(PANEL=NULL, FILE=NULL, Fusion.P=as.numeric(TWAS.P), Fusion.Z=as.numeric(TWAS.Z)) %>% mutate(GWAS=f, Region=cur.dir, Fusion.fdr=p.adjust(Fusion.P, method="fdr"))
  }
} # "" (skipped in TWAS) is converted to NA; " NA"  (failed in TWAS) is also converted to NA but generated a warning.

fusion.res = do.call(rbind, fusion.res)
rownames(fusion.res) = NULL

# missig GWAS...
setdiff(brainCatalog$files_names, gsub("_associations.txt", ".txt", fusion.res$GWAS)) # Arun: they do not have Beta/Odds ratio values

# check if GWAS has been run in all 4 regions.
if (all(with(fusion.res, table(GWAS, Region))>0)) {
  print("Fusion results were successfully read !")} else {
  print("Fusion results had errors !")
}
```

#### SMR
check folder name: "Fusion associations and SMR were completed for CAUC, while SMR was cancelled for AA. " 1/15/25 Arun

```{r}
smr.res = list()
for (cur.dir in c("DLPFC-CAUC", "mPFC-CAUC", "Hippo-CAUC", "Caudate-CAUC")) {
  cur.files = list.files(path = paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/SMR/smr_associations/"), pattern = ".smr$") %>% data.frame(filename=.) %>% mutate(ID = gsub(".smr", ".txt", filename)) %>% filter(ID %in% brainCatalog$files_names)
  
  for (f in cur.files$filename) {
    temp = read.table(paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/SMR/smr_associations/", f), header = T, sep = "\t")
    smr.res[[paste0(f, cur.dir)]] = temp %>% mutate(GWAS=f, Region=cur.dir, SMR.fdr=p.adjust(p_SMR, method="fdr"))
  }
}
length(smr.res) # 214, The SMR is performed on GWAS which has significant associations in Fusion analysis: Arun took all the significant TWAS associated traits and ran it across all the miRNAs for SMR analysis. 

smr.res = do.call(rbind, smr.res)
rownames(smr.res) = NULL

# check if GWAS has been run in all 4 regions.
all(with(smr.res, table(GWAS, Region))>0)
```

#### Merge
```{r}
fusion.res = mutate(fusion.res, GWAS=gsub("_associations.txt", "", GWAS))
smr.res = mutate(smr.res, GWAS=gsub(".smr", "", GWAS))

twas = full_join(fusion.res, smr.res, by=c("Region"="Region", "GWAS" = "GWAS", "ID"="Gene")) %>% dplyr::select(ID, GWAS, Region, everything())

nrow(twas) # 11118
with(twas, table(!is.na(Fusion.P), Region))
```

Filter NA values in GWAS
```{r}
#subset(twas, is.na(BEST.GWAS.ID))
#twas = subset(twas, !is.na(BEST.GWAS.ID))
#nrow(twas) # 9792
```

#### Visualization
```{r}
temp = twas %>% group_by(Region, GWAS) %>% slice_min(order_by = Fusion.P, n = 1) %>% ungroup()
p = ggplot(as.data.frame(temp), aes(x=GWAS, y=-log10(Fusion.P)))
p = p + geom_boxplot() + geom_point(aes(col=Region))
p + coord_flip()
```

# 2. Significant TWAS

### Fusion results for a given region
```{r}
cur.region = "DLPFC-CAUC"
temp = subset(twas, Fusion.fdr<0.05 & Region==cur.region)
temp = reshape2::dcast(subset(twas, Region==cur.region & ID %in% unique(temp$ID)), ID~GWAS, value.var = "Fusion.P") %>% tibble::column_to_rownames("ID")
temp = temp[, -which(apply(temp, 2, function(x) all(is.na(x))))] # remove gwas without any significant genes
dim(temp)
temp = -log10(temp)
temp[is.na(temp)] = -1
pheatmap::pheatmap(temp)
```

### Fusion & SMR
```{r}
twas.sig = subset(twas, Fusion.fdr<0.05 & p_SMR<0.05 & p_HEIDI>0.05)
nrow(twas.sig) # 68
length(unique(twas.sig$probeID)) # 18
```

Plot
```{r}
temp = twas.sig %>% group_by(GWAS) %>% summarise(ymin = min(Fusion.P), ymax = max(Fusion.P))
p = ggplot(twas.sig, aes(x=GWAS, y=-log10(Fusion.P)))
p = p + geom_point(aes(shape=Region), size=3)
p = p + geom_text(aes(label=probeID), vjust = -0.5)
p = p + geom_segment(data = temp, aes(x = GWAS, xend = GWAS, y = -log10(ymax), yend = -log10(ymin)), color = "blue", linewidth = 0.5)
p + coord_flip()

p = ggplot(twas.sig, aes(x=GWAS, y=-log10(Fusion.P)))
p = p + geom_point(aes(shape=Region, col=Region), size=3)
p = p + ggrepel::geom_text_repel(aes(label=probeID), size=2)
p = p + geom_segment(data = temp, aes(x = GWAS, xend = GWAS, y = -log10(ymax), yend = -log10(ymin)), color = "blue", linewidth = 0.5)
p = p + theme(axis.text.x=element_text(size=5, angle=90, hjust=1, vjust=1, color = "black"))
p
```

Examples
```{r}
subset(twas.sig, -log10(Fusion.P)>6)
subset(twas.sig, ID=="MI0008329_hsa-miR-1908-5p")
```

> MI0008329_hsa-miR-1908-5p (Nature aging)


```{r}
saveRDS(twas, "~/HwangLab/taeyoungh/miRNA/Results/twas_CAUC.rds")
```

# 3. Clean output
```{r}
# gwas information
#brainCatalog = readxl::read_xlsx("/mnt/Data1/HwangLab/taeyoungh/miRNA/Adult_Freeze-040625/1_Dataset/BrainCatalog_Metadata.xlsx")
#brainCatalog = subset(brainCatalog, trait_type == "binary")

temp = left_join(twas, mutate(brainCatalog, gwas_id=gsub(".txt", "", files_names)) %>% dplyr::select(gwas_id, pmid, year, n_total, n_cases, n_controls, trait_full_name), by=c("GWAS"="gwas_id")) # note that n_total can be different in the same pmid because some pmids have multiple traits.
nrow(temp) == nrow(twas) # confirm 1-to-1 map
twas = temp

# filter neuroticism (pmid: 29500382), social interaction (pmid: 29970889), sleep habit (30696823, 30846698), smoking (20418890), anthropometic (23563607), Method paper (34737426).
twas = subset(twas, !(pmid %in% c(29970889, 30696823, 30846698, 20418890, 23563607, 34737426)))

length(unique(twas$GWAS)) # 87

twas.sig = subset(twas, Fusion.fdr<0.05 & p_SMR<0.05 & p_HEIDI>0.05)
nrow(twas.sig) # 51
length(unique(twas.sig$probeID)) # 15
```

```{r}
saveRDS(twas, "~/HwangLab/taeyoungh/miRNA/Results/twas_CAUC_filtered.rds")
```


# 3. Examples

```{r}
subset(twas.sig, grepl("Schizophrenia", GWAS)) # MI0012063_hsa-miR-2682-5p
subset(subset(twas, Fusion.fdr<0.05), grepl("SCZ", GWAS)) # MI0012063_hsa-miR-2682-5p
```

> hsa-miR-2682-5p: a microRNA located within the MIR137HG (MIR137/MIR2682 locus). But it failed to pass HEIDI test (0.04) in SCZ_PGCwave3 DLPFC.

mir-137

```{r}
subset(twas, ID=="MI0000454_hsa-miR-137-3p")
```

```{r}
# heritability check
gcta_raw_files = list.files("DLPFC-CAUC/TWAS/GCTA/output_files/", pattern = ".*_h2.hsq", full.names = TRUE)
grep("hsa-miR-2682", gcta_raw_files, value=T)

mir137.h2 = list()

for (cur.dir in c("DLPFC-CAUC", "mPFC-CAUC", "Hippo-CAUC", "Caudate-CAUC", "DLPFC-AA", "mPFC-AA", "Hippo-AA", "Caudate-AA")) {
    temp = read.table(paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/GCTA/output_files/MI0000454_hsa-miR-137-3p_h2.hsq"), header = T, sep = "\t", fill = T)
    mir137.h2[[cur.dir]] = temp %>% mutate(Region=cur.dir)
}
mir137.h2 = do.call(rbind, mir137.h2)

subset(mir137.h2, Source=="Pval")
```

miR-499a-5p (Nature aging)
```{r}
# heritability check
grep("miR-499a", gcta_raw_files, value=T)

mir499a.h2 = list()

for (cur.dir in c("DLPFC-CAUC", "mPFC-CAUC", "Hippo-CAUC", "Caudate-CAUC", "DLPFC-AA", "mPFC-AA", "Hippo-AA", "Caudate-AA")) {
    temp = read.table(paste0("/mnt/Data1/HwangLab/Arun/adult_finalFreeze-082825/", cur.dir, "/TWAS/GCTA/output_files/MI0003183_hsa-miR-499a-5p_h2.hsq"), header = T, sep = "\t", fill = T)
    mir499a.h2[[cur.dir]] = temp %>% mutate(Region=cur.dir)
}
mir499a.h2 = do.call(rbind, mir499a.h2)

subset(mir499a.h2, Source=="Pval") # 0.017
```



```{r}
# expression check
miRcount.vst.lmCov5.res = readRDS("/home/taeyoungh/HwangLab/taeyoungh/miRNA/Adult_Freeze-040625/1_Data/miRcount.vst.lmCov5.res.rds")
sampleSheet = readRDS("/home/taeyoungh/HwangLab/taeyoungh/miRNA/Adult_Freeze-040625/1_Data/Freeze-040625_miRNA_qc_sampleSheet_20261010.rds")

temp = data.frame(t(miRcount.vst.lmCov5.res[c("hsa-miR-2682-5p", "hsa-miR-137-3p"),])) %>% tibble::rownames_to_column("RNum") %>% left_join(sampleSheet, by="RNum")
p = ggplot(temp, aes(x=hsa.miR.137.3p, y=hsa.miR.2682.5p, col=Race)) + facet_grid(PrimaryDx~BrainRegion)
p = p + geom_point()
p
```

```{r}
# eQTL check
temp.expr = read.table("DLPFC-CAUC/eQTL/dlpfc-cauc_miRNA_expr.bed", header=T, sep="\t", comment.char = "")

temp.genotype = bigsnpr::snp_attach("DLPFC-CAUC/eQTL/dlpfc-cauc_miRNA_tsqtl-cis-nominal.rds")

temp.eqtl = arrow::read_parquet("DLPFC-CAUC/eQTL/cis_nominal/dlpfc-cauc_miRNA_tsqtl-cis-nominal.cis_qtl_pairs.1.parquet") %>% left_join(temp.genotype$map[, c("marker.ID", "chromosome", "physical.pos", "allele1", "allele2")], by= c("variant_id" = "marker.ID")) %>% dplyr::rename(variant_chrom=chromosome, variant_pos=physical.pos)

subset(temp.eqtl, phenotype_id=="MI0012063_hsa-miR-2682-5p") %>% arrange(pval_nominal)
temp = subset(temp.eqtl, phenotype_id=="MI0012063_hsa-miR-2682-5p" & variant_id=="rs17117459")
plot.df = plot.eqtl.raw(temp, temp.expr, temp.genotype)
table(plot.df$genotype)
temp = subset(temp.eqtl, phenotype_id=="MI0000454_hsa-miR-137-3p" & variant_id=="rs17117459")
plot.df = plot.eqtl.raw(temp, temp.expr, temp.genotype)

subset(temp.eqtl, phenotype_id=="MI0000454_hsa-miR-137-3p") %>% arrange(pval_nominal, abs(start_distance))
temp = subset(temp.eqtl, phenotype_id=="MI0000454_hsa-miR-137-3p" & variant_id=="rs11165783")
plot.df = plot.eqtl.raw(temp, temp.expr, temp.genotype)
table(plot.df$genotype)
temp = subset(temp.eqtl, phenotype_id=="MI0012063_hsa-miR-2682-5p" & variant_id=="rs11165783")
plot.df = plot.eqtl.raw(temp, temp.expr, temp.genotype)
```
> minor allele causes different regulations between miR-137 and miR-2682 ?

```{r}
subset(temp.eqtl, phenotype_id=="MI0012063_hsa-miR-2682-5p" & pval_nominal<5*10^-6 & af>0.1 & abs(slope)>0.1) 

temp = subset(temp.eqtl, phenotype_id=="MI0012063_hsa-miR-2682-5p" & variant_id=="rs1702294")
plot.df = plot.eqtl.raw(temp, temp.expr, temp.genotype)
table(plot.df$genotype)
```

# 4. Tissue-specific TWAS

### miRNAs tested in all 4 brain regions
```{r}
temp = with(twas, table(ID, Region, GWAS))
temp = lapply(1:dim(temp)[3], function(i) { temp[, , i]}) # a list of 2 by 2 tables
temp = lapply(temp, function(x) which(rowSums(x)==4)) %>% unlist() %>% names() %>% table()
twas.gene.regionwide = names(temp)[temp == length(unique(twas$GWAS))] # 4: "MI0012063_hsa-miR-2682-5p" "MI0015873_hsa-miR-2355-3p" "MI0016005_hsa-miR-3615-3p" "MI0017320_hsa-miR-1343-3p"
```

Examples

```{r}
cur.id = "MI0017320_hsa-miR-1343-3p"
subset(twas, ID==cur.id & Fusion.fdr<0.05)

temp = reshape2::dcast(subset(twas, ID == cur.id & !is.na(Fusion.Z)), GWAS~Region, value.var = "Fusion.P")
temp = temp %>% tibble::column_to_rownames("GWAS")
pheatmap::pheatmap(-log10(temp))
```
MI0016005_hsa-miR-3615-3p: SCZ only in Hippo
MI0012063_hsa-miR-2682-5p: SCZ only in DLPFC and Caudate


